Support for one locale per socket.

jhh67 opened the pull request:
Adds the ability to run multiple locales per node as determined by the CHPL_RT_LOCALES_PER_NODE environment variable. The default is to run one locale per node. If the number of locales per node is less than or equal to the number of sockets, then each locale uses only the cores associated with its socket. 

This functionality is only supported for the following configuration:

CHPL_COMM = ofi
CHPL_LAUNCHER = slurm-srun
CHPL_COMM_OFI_OOB = pmi2
CHPL_TASKS = qthreads
CHPL_QTHREAD_TOPOLOGY = binders
CHPL_HWLOC = bundled

ronawho commented on runtime/src/comm/ofi/comm-ofi-oob-pmi2.c:
```diff
@@ -290,53 +290,64 @@ void decode_kvs(char* raw, const char* enc, size_t size) {
+  int       nodeID;   // locale's ID
+} locale_info_t;
+
+int chpl_comm_ofi_oob_locales_on_node(int *rank) {
```
Want to chat about this code -- seems like we should be able to get this information from PMI, but I think we're computing it ourselves now

ronawho replied:
Plan is to see if we can do this with PMI. May be some old context/info about how to do this in:
 - https://github.com/chapel-lang/chapel/pull/18279
 - https://github.com/chapel-lang/chapel/pull/18181

jhh67 replied:
Done.

ronawho commented on runtime/src/launch/slurm-srun/launch-slurm-srun.c:
```diff
@@ -240,6 +243,21 @@ static char* chpl_launch_create_command(int argc, char* argv[],
 
+  localesPerNode = 1;
+  if (lpnEnv != NULL) {
+    localesPerNode = atoi(lpnEnv);
```
I _think_ we can use `chpl_env_str_to_int` wrapper (or maybe can just use `chpl_env_rt_get_int`?)

jhh67 replied:
Changed it to use `chpl_env_rt_get_int`.

ronawho commented on runtime/src/launch/slurm-srun/launch-slurm-srun.c:
```diff
@@ -240,6 +243,21 @@ static char* chpl_launch_create_command(int argc, char* argv[],
+
+  if (localesPerNode > numLocales) {
+    localesPerNode = numLocales;
+  }
```
What is this adjustment doing?

ronawho replied:
This was for when running something like `CHPL_RT_LOCALES_PER_NODE=2 ./app -nl 1` to give that locale the entire node, but I think we've decided to back that out for now since we're not quite sure what the right default policy is here

jhh67 replied:
Removed this adjustment from the code so, for example, if you specify 2 locales per node and run only 1 locale it will only use the cores in its socket.

ronawho commented:
I'd like to see some performance numbers like in https://github.com/chapel-lang/chapel/pull/20875. 

I'm also curious what happens w/ and w/o a dedicated AM handler core. With a dedicated core I believe we'll restrict the thread to a core on the local socket, but without a dedicated core I'm wondering if we should still restrict the thread to the socket.

ronawho commented:
Oh, seems like we'll need to adjust the definition of `chpl_get_oversubscribed`. I think as-is we'll detect this as oversubscribed and we'll end up disabling qthreads affinity and lowering the worker thread spincount

jhh67 commented on runtime/src/tasks/qthreads/Makefile:
```diff
@@ -19,6 +19,14 @@
+ifdef DEBUG_NODEID
+RUNTIME_DEFS += -DDEBUG_NODEID
+endif
+
```
Move to https://github.com/chapel-lang/chapel/pull/21060.

jhh67 commented on runtime/src/topo/hwloc/topo-hwloc.c:
```diff
@@ -272,26 +270,74 @@ void getCPUInfo(void) {
+  // number of sockets and we get our own socket.
+  //
+
+  int numLocalesOnNode = chpl_get_num_locales_on_node();
```
Done.

jhh67 commented:
These are the performance results from running the core tests (https://github.com/Cray/chapel-private/issues/3593) on an HPE Cray EX.

| **Branch**         |      **main** |        **PR** |      **main** |       **main** |         **PR** |
|:-------------------|--------------:|--------------:|--------------:|---------------:|---------------:|
| **Nodes**          |             1 |             1 |             1 |             16 |             16 |
| **Locales**        |             1 |             2 |             2 |             16 |             32 |
| **Benchmark**      |               |               |               |                |                |
| stream             |       242.683 |       214.209 |       174.497 |        2693.55 |        3264.76 |
| stencil-opt        | 133958.802073 | 104712.373685 | 712121.603444 | 1515432.456987 | 1564049.310864 |
| isx-hand-optimized |       6.93509 |       4.79856 |       5.12267 |        7.37626 |         6.5307 |
| ig                 |         0.187 |         0.757 |         0.964 |          1.717 |          1.533 |
| ra-rmo             |      0.898138 |     0.0228166 |     0.0234886 |       0.211629 |       0.209251 |
| ra-on              |      0.605064 |    0.00440843 |    0.00571253 |       0.019994 |      0.0381767 |

Notes:

- The first column is the `main` branch with `-nl 1`. In this configuration the communication system is not initialized and the heap is  not pinned during memory registration. This means that the heap is faulted-in on demand, so there should be perfect memory affinity especially for the `stream` benchmark (each core will access memory only through its local memory controller). The second column is this PR with support for one locale per socket. In this configuration the memory system is initialized and the heap is pinned during memory registration. Although each locale will pin the heap to its socket, it will stripe the pages across the memory controllers. There is a 20% latency penalty on this machine when a core accesses memory through a different controller in the same socket. There are 4 memory controllers, which means that 75% of the memory accesses incur a 20% penalty, resulting in worse performance than the first column.
- The third column is the `main` branch with two locales running on a single node in a oversubscribed fashion. The communication system is initialized in this configuration.
- The last two columns show the benchmarks running on 16 nodes. For this PR there is one locale per socket, for a total of 32 locales. I'm not sure why `isx-hand-optimized` has worse performance than the `main` branch.

ronawho commented on runtime/src/topo/hwloc/topo-hwloc.c:
```diff
@@ -756,6 +832,11 @@ int chpl_topo_bindCPU(int id) {
+chpl_bool chpl_topo_isOversubscribed(void) {
+  CHK_ERR(initialized);
+  _DBG_P("oversubscribed = %s\n", oversubscribed ? "True" : "False");
+  return oversubscribed;
```
I think this new code is no longer respecting the `CHPL_RT_OVERSUBSCRIBED` env var

jhh67 replied:
Fixed.

ronawho commented on runtime/src/topo/hwloc/topo-hwloc.c:
```diff
@@ -281,26 +273,83 @@ void getCPUInfo(void) {
+        chpl_error(msg, 0, 0);
+      } else {
+        // We are oversubscribed.
+        oversubscribed = true;
```
I'm having a slow time parsing this code or at least following the logic of when we will or will not auto set oversubscribed. I think before we'd only auto-set it for ofi/ugni since I don't think we set numLocalesOnNode for gasnet. And here I think we'd restricting that so we won't set it if `LOCALES_PER_NODE = processOnNode` (which seems right) but then I'm wondering if this disables any other cases that don't current set rank?

jhh67 commented:
Modifications since last review:

* Changed oversubscription determination logic so that we are oversubscribed if 1) there is more than one locale per node and each doesn't have its own socket or 2) `CHPL_RT_OVERSUBSCRIBED` is true. Note that this logic is still too simplistic, instead we should determine if locales are sharing PUs, but that improvement will be in a subsequent PR.

* Get the entire system topology from `hwloc` and use that to determine if all PUs are accessible to the locale. By default `hwloc` gives us only accessible resources and we can't determine if there are inaccessible resources. We only want to limit ourself to a core if all resources are accessible, otherwise the launcher may have partitioned resources between tasks and we should simply use what we have been given. Note that this solves the problem under some `slurm` configurations in which if you run an odd number of locales and two locales per node, `slurm` binds the oddball locale to 32 cores on each socket, and if the locale subsequently limits itself to a single socket it ends up with only 32 cores whereas the rest of the locales have 64. With this change the oddball locale ends up with 64 cores (albeit across two sockets, I don't think there is any way I can fix that, see issue https://github.com/Cray/chapel-private/issues/4139).

* Removed the `--ntasks-per-node` option to `srun/sbatch`. It isn't needed if both `--ntasks` and `--nodes` are specified, and it causes a warning if the number of nodes doesn't divide evenly into the number of nodes.

* Determines the accessible NUMA domains correctly based on the accessible PUs.

* Added `lps_test.py` which contains locale-per-socket tests using the Python `unittest` framework.

ronawho commented on runtime/src/launch/slurm-srun/launch-slurm-srun.c:
```diff
@@ -89,18 +90,11 @@ static sbatchVersion determineSlurmVersion(void) {
   }
 }
 
-static int nomultithread(int batch) {
```
Don't think we can get rid of the nomultithread logic. This was added in https://github.com/chapel-lang/chapel/pull/15176 to address a user issue where a site had `SLURM_HINT=nomultithread` set and if we tried to ask for the logical cores it would fail to allocate nodes. See that PR and the linked issues for more details.

jhh67 replied:
The situation is rather complicated, as are all things related to `slurm`.  Keeping the `--hint=nomultithread` functionality as-is won't result in the desired behavior. If one runs 2 locales on an EX node with 2 sockets, 64 cores/socket, 2 PUs/core, and `--hint=nomultithread`, one would expect that each locale would run on its own socket using 64 PUs on 64 cores.  Instead, what's likely to happen with the current launcher implementation (at least under some `slurm` configurations) is that each locale will be bound to the same 128 PUs on the 32 cores in each socket (i.e., half of the cores in each socket go unused, the locales share PUs, and each locale spans both sockets). This is obviously undesirable. Here are some relevant things that cause this to happen. (I use the term `slurm` below to refer to both `srun` and `sbatch`).

* Note that the `--cpus-per-task` `slurm` option refers to PUs if hyperthreading is  enabled and cores if it is not.
* The Chapel `slurm-srun` launcher passes the `--exclusive` option to `slurm` so that Chapel jobs have exclusive access to nodes. This option also causes co-located tasks of a job use disjoint PUs (tasks don't share a PU). 

* The `slurm` `--cpu-bind=none` option doesn't mean that there is no CPU-binding under any circumstance, it means that the individual tasks of a job are not bound to disjoint PUs. If the tasks do not consume all PUs then they will be bound to their subset of the PUs. If `--cpu-bind` is not specified then each task will be bound to disjoint PUs. Here is a table to illustrate. There are 2 tasks (locales) on an EX and the `--exclusive `option is specified. Note that this is under a particular `slurm` configuration, other configurations may behave differently.

| --cpus-per-task | --cpu-bind | Result                                                         |
|-----------------|------------|----------------------------------------------------------------|
|             128 | N/A        | Each task bound to 128 PUs in its own socket                   |
|              64 | N/A        | Each task bound to 64 PUs on 32 cores in its own socket        |
|             128 | none       | No CPU binding at all                                          |
|              64 | none       | Each task bound to the same 128 PUs on 32 cores in each socket |

* In the first two rows `--cpu-bind` is not specified and `slurm` binds the tasks to disjoint PUs. If `--cpu-bind=none` then `slurm` either doesn't bind the tasks at all if `--cpus-per-task=128` because the tasks use all PUs on the node (row 3), or if `--cpus-per-task=64` it allocates a total of 128 PUs for the 2 tasks and binds both tasks to all 128 PUs (row 4).

* The `--hint=nomultithread` option tells `slurm` not to use `hyperthreading`  (i.e. only bind tasks to one PU per core). Note that the `--hint` and `--cpu-bind` options are mutually exclusive; if both are specified `slurm` ignores the `--hint` option (the `SLURM_HINT` environment variable has the same behavior except that `slurm` doesn't print a warning about the hint being ignored). This makes sense because` --hint=nomultithread` tells `slurm` not to bind tasks to more than one PU per core, while `--cpu-bind=none` tells `slurm` not to bind the tasks at all. The latter takes precedence). Here is the same table as above with `--hint=nomultithread` specified.

| --cpus-per-task | --cpu-bind | Result                                                                        |
|-----------------|------------|-------------------------------------------------------------------------------|
|             128 | N/A        | error: Requested node configuration is not available                          |
|              64 | N/A        | Each task bound to 64 cores in its own socket                                 |
|             128 | none       | No CPU binding at all (hint ignored)                                          |
|              64 | none       | Each task bound to the same 128 PUs on 32 cores in each socket (hint ignored) |

* Row 1 is an error because `slurm` is asked to run 2 tasks on 128 disjoint PUs without using hyperthreading, which is impossible.

* The current implementation of the Chapel `slurm-srun` launcher does not  specify `--cpu-bind=none` and if `--hint=nomultithread` it divides the number of CPUs per task by the number of PUs/core to calculate `--cpus-per-task`. This corresponds to row 2, so that each task is bound to the cores in its own socket (note again that different slurm configurations may behave differently).

* There is at least one customer system in which there is a system-wide  `SLURM_HINT=nomultithread` to essentially disable hyperthreading.

* The locale-per-socket functionality requires` --cpu-bind=none` so that each locale can limit itself to its own socket.  However, `--cpu-bind=none` will cause slurm to ignore `--hint=nomultithread`. As currently implemented,  the launcher will still divide by the number of PUs/core to end up with `--cpus-per-task=64` (row 4 above) which is not the outcome we want.

* The reason for specifying `--hint=nomultithread` is to prevent the use  of hyperthreading. Note that by default the Chapel runtime does not use hyperthreading; it creates one qthread shepherd per core. So it does the "right thing" whether or not `--hint=nomultithread` is specified. This corresponds to row 3 above in which `--cpus-per-task=128`, `--cpu-bind=none`, and `--hint=nomultithread` is ignored.

* The user can force Chapel to use hyperthreading by setting `CHPL_RT_NUM_THREADS_PER_LOCALE` to a value greater than the number of cores.

* Base on all of these considerations, if `--hint=nomultithread` the proper behavior by the launcher is to:

1.  **not** reduce `--cpus-per-task`, and
2. limit `CHPL_RT_NUM_THREADS_PER_LOCALE` to the number of cores so that the Chapel runtime does not use hyperthreading.

jhh67 replied:
I've done a few experiments since the above comment and have an alternative solution:

1. *Not* reduce `--cpus-per-task`, same as above.
2. Always specify `--cpu-bind=none` even if there is one locale per node. If you don't specify this then #1 will cause an error because `srun` can't bind the locale to that many cores.
3. If `--hint=nomultithread` have the launcher set an environment variable (e.g. `CHPL_RT_NO_MULTITHREAD`). The runtime uses this in `setupAvailableParallelism` in `tasks-qthreads.c` to determine whether to limit the available parallelism to logical or physical CPUs.

jhh67 replied:
I implemented a slightly different alternative from above that doesn't change the behavior when there is a single locale per node.

1. Only reduce `--cpus-per-task` if there is a single locale per node.
2. Only specify `--cpu-bind=none` if there is more than one locale per node.
3. If there is more than one locale per node and `--hint=nomultithread`, the launcher sets the environment variable `CHPL_RT_NO_MULTITHREAD` that the runtime uses to determine whether to limit the available parallelism to logical or physical CPUs.

ronawho commented on runtime/src/topo/hwloc/topo-hwloc.c:
```diff
@@ -144,12 +158,18 @@ void chpl_topo_init(void) {
+  // See what we are missing.
+  int flags = HWLOC_TOPOLOGY_FLAG_WHOLE_SYSTEM;
+  CHK_ERR_ERRNO(hwloc_topology_set_flags(wholeTopology, flags) == 0);
+  CHK_ERR_ERRNO(hwloc_topology_load(wholeTopology) == 0);
```
Loading a topology is relatively expensive (not really in the context of distributed set up, but at least for comm=none it adds a non-trivial amount of time to load a topology.) This is why we go out of our way to give qthreads our topology instead of having qthreads also load. For now I think we were talking about conditionalizing this on comm!=none

For more info on hwloc load time see https://github.com/chapel-lang/chapel/issues/7534 and https://github.com/Qthreads/qthreads/issues/41 and

jhh67 replied:
I modified the implementation to use the `online_cpuset` of the topology rather than getting the topology for the whole system.

jhh67 commented on runtime/src/topo/hwloc/topo-hwloc.c:
```diff
@@ -311,6 +312,9 @@ void getCPUInfo(void) {
                                                          pu));
+    int numPus = hwloc_bitmap_weight(core->cpuset);
+    CHK_ERR((pusPerCore == 0) || (pusPerCore == numPus));
+    pusPerCore = numPus;
```
See the in-line comment below, but there isn't a good way to get the number of cores from the online cpuset because PUs and cores that are online but inaccessible don't have representative objects in the topology. So you can't call `hwloc_get_nbobjs_inside_cpuset_by_type(...,HWLOC_OBJ_CORE...)` for example. I make the assumption that all cores have the same number of PUs and compute the number of cores from the numbers of PUs.

ronawho reviewed (approved):
Code looks good, but can you regather perf figures so we have a record of where things ended up. Don't need the full table, just 16 node results for:
 - main with 1 LPN
 - PR with 1 LPN
 - PR with 2 LPN

jhh67 commented:
Here are the final performance numbers for this PR as measured on a HPE Cray EX:

| **Branch**         | **main**  | **PR**    | **PR**    |
|--------------------|-----------|-----------|-----------|
| **Nodes**          |        16 |        16 |        16 |
| **Locales**        |        16 |        16 |        32 |
| **Benchmark**      |           |           |           |
|             stream |   2695.36 |   2689.45 |   3262.74 |
|        stencil-opt |   1488610 |   1523650 |   1580138 |
| isx-hand-optimized |   7.46517 |   7.44247 |   6.54057 |
|                 ig |     1.750 |     1.721 |     1.507 |
|             ra-rmo |  0.216066 |  0.218208 |   0.20594 |
|              ra-on | 0.0195256 | 0.0217225 | 0.0337504 |
